homeassistant/components/telegram_bot/webhooks.py

```diff
@@ -36,16 +36,18 @@ async def async_setup_platform(hass, bot, config):
         _LOGGER.error("Invalid telegram webhook %s must be https", pushbot.webhook_url)
         return False
 
+    await pushbot.start_application()
     webhook_registered = await pushbot.register_webhook()
     if not webhook_registered:
         return False
 
     hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP, pushbot.deregister_webhook)
+    hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP, pushbot.stop_application)
```

Comment:
Maybe register a single handler for the stop event that deregisters the webhook and stops the application?

Reply:
I'm not sure why, and I've spent quite a bit of time on it, but for some reason the mock doesn't work when the `deregister_webhook` is called from within the other method. I'm guessing something to do with the patch context manager clsoing before the signal, but that doesn't explain why having them separate still works 🤷

Reply:
That being said, it's not really a great idea to have application code a certain way because of tests

Reply:
Please specify which test fails.

Reply:
I don't have access to my dev machine right now, but I believe it was any that uses the `webhook_platform` fixture. 

When the test is tearing down it runs the `stop_application`, which runs the `deregister_webook`, which fails on 401 because it calls telegram API with a bad token/botID. When `hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP, pushbot.deregister_webhook)` the mock seems to work, preventing the API call.

Reply:
In `webhook_platform` we can `yield` after setting up the integration and then call await `hass.async_stop()` to fire the event while we still have the bot patch in place.

Reply:
that worked perfectly, thank you again for sharing your test prowess